DOCU-2377: Cleanup KIC using kongplugin resource guide (#4064)
Kong/docs.konghq.com: src/kubernetes-ingress-controller/guides/using-kongplugin-resource.md

title: Using KongPlugin resource
---

> This guide will cover how to use the {{site.kic_product_name}} 

Reviewer: ```suggestion
This guide walks you through using the {{site.kic_product_name}} 
```

[deployment guides](/kubernetes-ingress-controller/{{page.kong_version}}/deployment/overview) to configure this environment variable.

> If everything is setup correctly, making a request to Kong should return

Reviewer: ```suggestion
If everything is set up correctly, making a request to Kong should return
```

`HTTP 404 Not Found`.

```sh
curl -i $PROXY_IP
```

> In this document, expected output will follow each command:

Reviewer: ```suggestion
In this document, the expected output follows each command:
```

```sh
HTTP/1.1 404 Not Found
Content-Type: application/json; charset=utf-8
Connection: keep-alive
Content-Length: 48
X-Kong-Response-Latency: 1
Server: kong/2.8.1

{"message":"no Route matched with those values"}
```

This message is expected as Kong does not yet know how to proxy the request.

## Installing sample services

> We will start by installing two services. First an httpbin service:

Reviewer: ```suggestion
Start by installing two services. First, install an `httpbin` service:
```

```sh
kubectl apply -f https://bit.ly/k8s-httpbin
```

```sh
service/httpbin created
deployment.apps/httpbin created
```

> And then an echo service:

Reviewer: ```suggestion
And then an `echo` service:
```

## Installing sample services

We will start by installing two services. First an httpbin service:

```sh
kubectl apply -f https://bit.ly/k8s-httpbin
```

> ```sh

Reviewer: Can you preface each output block with "Expected output:"?

Author: I did that at first, and thought it was unnecessarily redundant.  If you feel strongly I can add it back in

Reviewer: Hmm. I see what you mean. Let's go with your way for now, and we can always revisit if people continue to run into issues.

If we send a request to `/baz`, then we can see that the header is not injected
by Kong:

```sh
curl -I $PROXY_IP/baz
```

```sh
HTTP/1.1 200 OK
Content-Type: text/html; charset=utf-8
Content-Length: 9593
Connection: keep-alive
Server: gunicorn/19.9.0
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
X-Kong-Upstream-Latency: 13
X-Kong-Proxy-Latency: 1
Via: kong/2.8.1
```

> Here, we have successfully setup a plugin which is executed only when a

Reviewer: ```suggestion
Here, we have successfully set up a plugin which is executed only when a
```

> Use the API key to pass authentication:

Reviewer: ```suggestion
Use the API key to pass authentication. Try it with `/baz`:
```

X-Kong-Upstream-Latency: 2
X-Kong-Proxy-Latency: 1
Via: kong/1.2.1
```

> ```sh

Reviewer: I wonder if having the exact say command tested on both `foo` and `baz` is useful here, or if it just creates clutter. If you think it makes sense to leave it, then preface it with:
```suggestion
Then use the API key with `/foo`:
```sh
```

Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
demo:  injected-by-kong
X-Kong-Upstream-Latency: 2
X-Kong-Proxy-Latency: 0
Via: kong/2.8.1
```

## Configure a global plugin

Instead of applying plugins to specific services or ingress routes,
we can apply plugins to protect the entire gateway.

> To show this, we will be configuring a rate-limiting plugin, which
> will throttle requests coming from the same client.

Reviewer: ```suggestion
To test this, let's configure a rate-limiting plugin to throttle requests coming from the same client.
```
Trying to get rid of the future tense whenever possible (and "we", but that's a whole other issue). I know these guides are full of it, so it's a slow change.

  policy: local
plugin: rate-limiting
" | kubectl apply -f -
```

```sh
kongclusterplugin.configuration.konghq.com/global-rate-limit created
```

With this plugin (please note the `global` label), every request through

> the {{site.kic_product_name}} will be rate-limited:

Reviewer: ```suggestion
the {{site.kic_product_name}} is rate-limited:
```

X-Kong-Upstream-Latency: 3
X-Kong-Proxy-Latency: 1
Via: kong/1.2.1
```

> ```sh

Reviewer: ```suggestion

Requests to `/bar` are also rate limited:
```sh
```

username: harry
credentials:
- harry-apikey" | kubectl apply -f -
```

```
kongconsumer.configuration.konghq.com/harry configured
```

Note the annotation being added to the `KongConsumer` resource.

Now, if the request is made as the `harry` consumer, the client
is rate-limited differently:

```sh
curl -I $PROXY_IP/foo -H 'apikey: my-sooper-secret-key'
```

```
HTTP/1.1 200 OK
Content-Type: text/html; charset=utf-8
Content-Length: 9593
Connection: keep-alive
X-RateLimit-Remaining-Minute: 9
X-RateLimit-Limit-Minute: 10
RateLimit-Remaining: 9
RateLimit-Reset: 42
RateLimit-Limit: 10
Server: gunicorn/19.9.0
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
demo:  injected-by-kong
X-Kong-Upstream-Latency: 2
X-Kong-Proxy-Latency: 1
Via: kong/2.8.1
```

And a regular unauthenticated request:

```sh
curl -I $PROXY_IP/bar
```

```sh
HTTP/1.1 200 OK
Content-Type: text/plain; charset=UTF-8
Connection: keep-alive
X-RateLimit-Remaining-Minute: 4
X-RateLimit-Limit-Minute: 5
RateLimit-Remaining: 4
RateLimit-Reset: 11
RateLimit-Limit: 5
Server: echoserver
demo:  injected-by-kong
X-Kong-Upstream-Latency: 1
X-Kong-Proxy-Latency: 1
Via: kong/2.8.1
```

> There is a lot more you can do with Kong Plugins, check the [Plugin Hub](/hub) for
> more information.
> 
> Next, you might want to learn more about Ingress with the 
> [KongIngress resource guide](/kubernetes-ingress-controller/{{page.kong_version}}/guides/using-kongingress-resource/).

Reviewer: ```suggestion
## Next steps

There's a lot more you can do with Kong plugins. Check the [Plugin Hub](/hub) to see all of the available plugins and how to use them.

Next, you might want to learn more about Ingress with the 
[KongIngress resource guide](/kubernetes-ingress-controller/{{page.kong_version}}/guides/using-kongingress-resource/).
```

Minor phrasing edit, mainly wanted to add the header. Also, contractions are totally fine to use (preferred, even).

Now, if the request is made as the `harry` consumer, the client

> will be rate-limited differently:

Reviewer: ```suggestion
is rate-limited differently:
```

```sh
curl -I $PROXY_IP/foo -H 'apikey: my-sooper-secret-key'
```

```
HTTP/1.1 200 OK
Content-Type: text/html; charset=utf-8
Content-Length: 9593
Connection: keep-alive
X-RateLimit-Remaining-Minute: 9
X-RateLimit-Limit-Minute: 10
RateLimit-Remaining: 9
RateLimit-Reset: 42
RateLimit-Limit: 10
Server: gunicorn/19.9.0
Access-Control-Allow-Origin: *
Access-Control-Allow-Credentials: true
demo:  injected-by-kong
X-Kong-Upstream-Latency: 2
X-Kong-Proxy-Latency: 1
Via: kong/2.8.1
```

> And a regular unauthenticated request

Reviewer: ```suggestion
And a regular unauthenticated request:
```

```sh
curl -I $PROXY_IP/bar
```

```sh
HTTP/1.1 200 OK
Content-Type: text/plain; charset=UTF-8
Connection: keep-alive
X-RateLimit-Remaining-Minute: 4
X-RateLimit-Limit-Minute: 5
RateLimit-Remaining: 4
RateLimit-Reset: 11
RateLimit-Limit: 5
Server: echoserver
demo:  injected-by-kong
X-Kong-Upstream-Latency: 1
X-Kong-Proxy-Latency: 1
Via: kong/2.8.1
```

> This guide demonstrates how you can use the {{site.kic_product_name}} to

Reviewer: This last statement seems super out of place at the end of the page. I know this isn't part of what you were fixing, but can you move it to the start of the topic instead? Otherwise, I'm not sure we'll ever remember to come back and fix it.

Author: Thanks for your help @lena-larionova 😄 , please check the latest where I modified the outro a bit, and I moved some of this information to the intro and tried to simplify and clarify it.

Also, I added a link to Plugin Hub, if you wouldn't mind double checking how I referenced that in the outro / CTA here on this page.  👍
